### DISCUSSION TITLE
Extending NMEA-Types

### DISCUSSION THREAD

**Discussion by chrissli** [Knio/pynmea2#146] [Q&A] [answered]
I'm sorry if this is a dumb question, but I fairly new to python.
I have a NMEA-ish string, I guess it would be a proprietary sentence, maybe.
However, what is the appropriate way to implement this here without editing the existent code?
I tried the following, but I wouldn't work:
---
import pynmea2
from pynmea2 import nmea

class pynmea2_XBO(pynmea2.talker.TalkerSentence):	
	fields = (
		('Temperature(outside)','temperature'),
		('Pressure','pressure'),
		('Humidity','humidity')		
	)
	
pynmea2.XBO = pynmea2_XBO

line = 'PXBO,23.3,992.2,56' # this is what the my strings are like

nmeaObj = pynmea2.parse(line)

print(nmeaObj)	
	#> $PXBO,23.3,992.2,56*18
print(vars(nmeaObj))
	#> {'manufacturer': 'XBO', 'data': ['', '23.3', '992.2', '56']}
---
		
Where the "fields" attribute exists but is empty.
Can you please provide an example how to inject own NMEA-sentences without touching your code?
Also, sorry for poorly formated code, I just can't get this textbox working

**Comment by Knio** [comment, marked as answer]
You were very close.
* you need to inherit from `ProprietarySentence`
* your class name *must* be `XBO`, this is what matches on the `PXBO` string
* you need an extra unused field at the beginning - this is because some proprietary sentences have an extra comma and some don't
* no need to assign `pynmea2.XBO` (unless you want to), just defining the class is enough
* `vars()` will not show the fields, since they're dynamically accessed, but `repr()` will

```python
import pynmea2

class XBO(pynmea2.nmea.ProprietarySentence):
    fields = (
        ("Blank", "_blank"),
        ('Temperature(outside)','temperature'),
        ('Pressure','pressure'),
        ('Humidity','humidity')
    )

line = 'PXBO,23.3,992.2,56'
nmeaObj = pynmea2.parse(line)

print(f'obj: {nmeaObj}')
print(f'type: {type(nmeaObj)}')
print(f'vars: {vars(nmeaObj)}')
print(f'repr: {repr(nmeaObj)}')
print(f'temperature: {nmeaObj.temperature}')
print(f'pressure: {nmeaObj.pressure}')
print(f'humidity: {nmeaObj.humidity}')
```

output:
```
obj: $PXBO,23.3,992.2,56*18
type: <class '__main__.XBO'>
vars: {'manufacturer': 'XBO', 'data': ['', '23.3', '992.2', '56']}
repr: <XBO(_blank='', temperature='23.3', pressure='992.2', humidity='56')>
temperature: 23.3
pressure: 992.2
humidity: 56
```

**Reply by chrissli** [reply]
Sorry, took me a bit to test it out. That works like a charm, thank you!

The reason why I assigned 
```
pynmea2.XBO = pynmea2_XBO
```

was, that I already got a class "XBO", so I had my hopes up that I can just assign the "real" class to the XBO attribute.
But that's not big of a deal, I'm just gonna adjust to that.
Again, thank you very much for your help.

**Reply by chrissli** [reply]
Wait no, I'm just dumb:
```
class XBOWhatever(pynmea2.nmea.ProprietarySentence):
    fields = (
        ("Blank", "_blank"),
        ('Temperature(outside)','temperature'),
        ('Pressure','pressure'),
        ('Humidity','humidity')
    )
pynmea2.ProprietarySentence.sentence_types['XBO'] = XBOWhatever
```
Will work without touching my existing class.
